### DISCUSSION TITLE
Guidance for creating new trainable components

### DISCUSSION THREAD

**Discussion by salmenf** [explosion/spaCy#6361] [Help: Coding & Implementations] [answered]
I've been following the guide "Create new trainable components" from the docs of spaCy 3 (nightly). 

Basically, I'm trying to hook in a `tf.keras.Model` to recognize custom entities, which should use vectors from a transformer as an input. My understanding: I need to build a `TrainablePipe` which uses a model that is just a simple `thinc.api.chain` using the Transformer as a tok2vec layer, and the `tf.keras Model` as the "main layer". The `TrainablePipe` should then implement the methods from the guide to serve as the spaCy interface to the model, allowing configuration via .cfg files, training via the CLI, and usage in pipelines.

I'm having trouble understanding the "initialize" method. The intent is to do shape inference based on a representative slice of training data, right? Does this work with a wrapped `tf.keras.Model`? And how do I pick a good slice, by the way?
Specifically, the line `self._examples_to_truth(subbatch)` confuses me. Where does this method come from? If it's a dummy not shown in the example, what should it do?
Also, the example is slightly incomplete without an implementation of `TrainablePipe.add_label`.

Some guidance on this would be great for me since I'm struggling with Deep Learning, to be honest. And maybe a sentence or two explaining that method would be good for the docs, also adding a code sample for `add_label`.

## Which page or section is this issue related to?

https://nightly.spacy.io/usage/layers-architectures#component-rel-pipe

**Comment by svlandeg** [comment, marked as answer]
Great to see you've been taking the `nightly` for a spin! Let me try and clarify some points.

- The `add_label` method is documented here: https://nightly.spacy.io/api/pipe#add_label - but its actual implementation depends on the component's internal fields. For instance, for textcat, it mainly does
```
self.cfg["labels"].append(label)
self.vocab.strings.add(label)
```
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; but you can see the full implementation (as an example) [here](https://github.com/explosion/spaCy/blob/develop/spacy/pipeline/textcat.py#L306). 

- The `_examples_to_truth` method is not a standard function, but we do use this name internally for some of our components, again you can have a look at the code in textcat as an example [here](https://github.com/explosion/spaCy/blob/develop/spacy/pipeline/textcat.py#L274). How this function looks like, depends on how the gold-standard data is encoded in the `Example` objects. For instance, the textcat goes through `eg.reference.cats`, which are the category annotations in the `reference` (gold-standard) document of the example `eg`. Then it converts those annotations to a `truths` matrix where each row `i` is an example, and each column `j` refers to one of the potential textcat labels stored in `self.labels`.

- The `initialize` method is documented [here](https://thinc.ai/docs/usage-models#validation) and [here](https://nightly.spacy.io/usage/training#initialization). The [shape inference](https://nightly.spacy.io/usage/layers-architectures#thinc-shape-inference) will only work with Thinc models AFAIK. To show you just one example, Thinc's linear layer, when [initialized](https://github.com/explosion/thinc/blob/master/thinc/layers/linear.py#L46), will set its input and output dimension according to the provided data sample, and will then initialize its internal `W` and `b` matrices.

- What a representative sample is, depends on the specifics of the task. For instance, the textcat can deduce the gold labels from the given sample data. But if at that point, you only provide examples with 3 different labels, and you afterwards try training it with new labels, that will fail. 

With respect to this section in the docs you linked: this is still a work in progress, and we'll soon be releasing the full implementation as an example [project](https://github.com/explosion/projects) to support the docs. We might even release a tutorial video of this to walk through the different steps in-depth ;-)

**Comment by salmenf** [comment]
Thanks a lot for your clarifications, and the quick reply! Your links and and explanations answered my questions perfectly. The documentation is impressively detailed, which is great for the most part, but in this case, I needed some guidance to navigate it, it seems. :) 

The nightly build is great so far, by the way, it made a lot of concepts accessible for me with the very detailed documentation. I appreciate it!
